dartagnan/src/main/java/com/dat3m/dartagnan/analysis/graphRefinement/GraphRefinement.java

```diff
@@ -0,0 +1,474 @@
+
+    // ============================================
+
+    // =============== Accessors =================
+
+    public VerificationTask getTask() {
+        return task;
+    }
+
+    // NOTE: The execution graph should not be modified from the outside!
+    public ExecutionGraph getExecutionGraph() { return execGraph; }
+
+    public ExecutionModel getCurrentModel() {
+        return executionModel;
+    }
+
+
+    // =============================================
+
+    // =========== Construction & Init ==============
+
+    public GraphRefinement(VerificationTask task) {
+        this.task = task;
+        this.execGraph = new ExecutionGraph(task);
+        this.executionModel = new ExecutionModel(task);
+        this.reasoner = new Reasoner(execGraph, true);
+    }
+
+    // ----------------------------------------------
+
+    private void populateFromModel(Model model, SolverContext ctx) {
+        executionModel.initialize(model, ctx, false);
+        execGraph.initializeFromModel(executionModel);
+        if (DEBUG) {
+            // TODO: Remove testing code
+            testIteration();
+            testStaticGraphs();
+        }
+    }
+
+    private List<Edge> createCoSearchList() {
+        Relation co = task.getMemoryModel().getRelationRepository().getRelation(CO);
+        Map<BigInteger, Set<Edge>> possibleCoEdges = new HashMap<>();
+        List<Edge> initCoherences = new ArrayList<>();
+        TupleSet minSet = co.getMinTupleSet();
+        TupleSet maxSet = co.getMaxTupleSet();
+
+        for (Map.Entry<BigInteger, Set<EventData>> addressedWrites : executionModel.getAddressWritesMap().entrySet()) {
+            Set<EventData> writes = addressedWrites.getValue();
+            BigInteger address = addressedWrites.getKey();
+            Set<Edge> coEdges = new HashSet<>();
+            possibleCoEdges.put(address, coEdges);
+
+            for (EventData e1 : writes) {
+                for (EventData e2: writes) {
+                    if (e1 == e2) {
+                        continue;
+                    }
+
+                    Tuple t = new Tuple(e1.getEvent(), e2.getEvent());
+
+                    if (!maxSet.contains(t)) {
+                        // The co(e1, e2) edge can never be contained in any execution
+                        // so co(e2, e1) must be present instead.
+                        initCoherences.add(new Edge(e2, e1));
+                        continue;
+                    } else if (minSet.contains(t)) {
+                        // Min-Set coherences have to be contained in all executions
+                        initCoherences.add(new Edge(e1, e2));
+                        continue;
+                    }
+
+                    // We only add edges in one direction since the search procedure will test
+                    // each coherence in both directions anyways!
+                    if (e2.getId() >= e1.getId()) {
+                        continue;
+                    }
```

Comment:
I guess you refer to iterating such that this condition is always satisfied. There are two reasons why that doesn't work:
- I have a ```Set``` and no ```List```. For what you want to do, you need ```ListIterator``` instead of ```Iterator```.
- The actual reason, however, is that I need both directions for the previous code that checks the max and min sets. That is why this check is performed afterwards.